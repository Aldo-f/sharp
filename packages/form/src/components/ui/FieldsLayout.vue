<template>
    <Grid :rows="layout">
        <template slot-scope="fieldLayout">
            <slot v-if="!fieldLayout.legend" v-bind="fieldLayout"></slot>

            <fieldset class="SharpForm__fieldset" v-else v-show="isFieldsetVisible(fieldLayout)">
                <div class="SharpModule__inner">
                    <div class="SharpModule__header">
                        <legend class="SharpModule__title">{{fieldLayout.legend}}</legend>
                    </div>
                    <div class="SharpModule__content">
                        <FieldsLayout :layout="fieldLayout.fields">
                            <template slot-scope="fieldset">
                                <slot v-bind="fieldset"></slot>
                            </template>
                        </FieldsLayout>
                    </div>
                </div>
            </fieldset>
        </template>
    </Grid>
</template>

<script>
    import { Grid } from 'sharp-ui';

    export default {
        name:'SharpFieldsLayout',

        components: {
            Grid
        },

        props: {
            layout: { // 2D array fields [ligne][col]
                type: Array,
                required: true
            },
            visible : {
                type: Object,
                default: () => ({})
            }
        },

        data() {
            return {
                fieldsetMap: {}
            }
        },

        methods: {
            isFieldsetVisible(fieldLayout) {
                let { id, fields } = fieldLayout;

                let map = this.fieldsetMap[id] || (this.fieldsetMap[id] = [].concat.apply([],fields));
                return map.some(f => this.visible[f.key]);
            }
        }
    }
</script>