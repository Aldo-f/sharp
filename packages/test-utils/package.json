{
    "name": "sharp-test-utils",
    "private": true
}