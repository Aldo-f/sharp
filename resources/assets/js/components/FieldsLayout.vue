<template>
    <sharp-grid :rows="layout">
        <template scope="field">
            <slot v-if="!field.fieldset" v-bind="field"></slot>
            <fieldset v-else>
                <legend>{{field.fieldset}}</legend>
                <sharp-fields-layout :layout="field.fields">
                    <template scope="fieldset">
                        <slot v-bind="fieldset"></slot>
                    </template>
                </sharp-fields-layout>
            </fieldset>
        </template>
    </sharp-grid>
</template>

<script>
    import Grid from './Grid';
    import FieldContainer from './FieldContainer';
    export default {
        name:'SharpFieldsLayout',

        components: {
            [FieldContainer.name]:FieldContainer,
            [Grid.name]:Grid
        },

        props: {
            layout: { // 2D array fields [ligne][col]
                type: Array,
                required: true
            }
        }
    }
</script>