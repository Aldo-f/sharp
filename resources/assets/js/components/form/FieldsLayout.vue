<template>
    <sharp-grid :rows="layout">
        <template scope="fieldLayout">
            <slot v-if="!fieldLayout.legend" v-bind="fieldLayout"></slot>
            <fieldset v-else>
                <legend>{{fieldLayout.legend}}</legend>
                <sharp-fields-layout :layout="fieldLayout.fields">
                    <template scope="fieldset">
                        <slot v-bind="fieldset"></slot>
                    </template>
                </sharp-fields-layout>
            </fieldset>
        </template>
    </sharp-grid>
</template>

<script>
    import Grid from '../Grid';

    export default {
        name:'SharpFieldsLayout',

        components: {
            [Grid.name]:Grid
        },

        props: {
            layout: { // 2D array fields [ligne][col]
                type: Array,
                required: true
            }
        }
    }
</script>