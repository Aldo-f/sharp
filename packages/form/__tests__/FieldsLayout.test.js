import FieldsLayout from '../src/components/ui/FieldsLayout.vue';

import { shallowMount } from '@vue/test-utils';

describe('fields-layout', () => {

    const FieldMock = { name:'Field', template:'<div class="FIELD_MOCK"></div>' };
    function createWrapper({ propsData }) {
        return shallowMount(FieldsLayout, {
            propsData: {
                ...propsData,
            },
            // language=Vue
            stubs: {
                Grid:
                    `<div class="GRID_MOCK">
                        <slot v-bind="rows[0][0]"></slot>
                    </div>`,
                FieldsLayout:
                    // render first field from given layout
                    `<div class="FIELDS_LAYOUT_MOCK">
                        <slot v-bind="$attrs.layout[0][0]"></slot>
                    </div>`,
                Field: FieldMock,
            },
            scopedSlots: {
                default: '<Field :data="props" />',
            },
        });
    }

    test('can mount fields layout', () => {
        const wrapper = createWrapper({
            propsData: {
                layout: [
                    [{ key:'title' }]
                ]
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount "fieldset" fields layout', () => {
        const wrapper = createWrapper({
            propsData: {
                layout: [
                    [{
                        legend: 'Fieldset', fields: [[
                            { key: 'title' }
                        ]]
                    }]
                ],
                visible: {
                    title: true
                }
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('can mount "hidden fieldset" fields layout', () => {
        const wrapper = createWrapper({
            propsData: {
                layout: [
                    [{
                        legend: 'Fieldset', fields: [
                            [{ key: 'title' }]
                        ]
                    }]
                ],
                visible: {
                    title: false
                }
            }
        });

        expect(wrapper.html()).toMatchSnapshot();
    });

    test('expose correct props', () => {
        const wrapper = createWrapper({
            propsData: {
                layout: [
                    [{ key:'list' }]
                ]
            }
        });

        expect(wrapper.find({ name: 'Field' }).vm.$attrs).toEqual({
            data: { key: 'list' }
        })
    });

    test('expose correct fieldset props', () => {
        const wrapper = createWrapper({
            propsData: {
                layout: [
                    [{
                        legend: 'Fieldset', fields: [
                            [{ key: 'title' }]
                        ]
                    }]
                ],
            }
        });

        expect(wrapper.find({ name: 'Field' }).vm.$attrs).toEqual({
            data: { key: 'title' }
        });
    });

    test('fieldset visible', () => {
        const wrapper = createWrapper({
            propsData: {
                layout: [
                    [{
                        legend: 'Fieldset 1',
                        id: 'fieldset_1',
                        fields: [
                            [{ key: 'title' }, { key: 'subtitle' }],[{ key: 'name' }]
                        ]
                    }]
                ],
                visible: {
                    title: true, subtitle: true, name: false,
                }
            }
        });

        expect(wrapper.vm.fieldsetMap).toEqual({
            'fieldset_1': [{ key: 'title' }, { key: 'subtitle' },{ key: 'name' }]
        });

        expect(wrapper.vm.isFieldsetVisible({ id: 'fieldset_1'})).toBe(true);
    });

    test('fieldset invisible', () => {
        const wrapper = createWrapper({
            propsData: {
                layout: [
                    [{
                        legend: 'Fieldset 1',
                        id: 'fieldset_1',
                        fields: [
                            [{ key: 'title' }, { key: 'subtitle' }],[{ key: 'name' }]
                        ]
                    }]
                ],
                visible: {
                    title: false, subtitle: false, name: false,
                }
            }
        });

        expect(wrapper.vm.fieldsetMap).toEqual({
            'fieldset_1': [{ key: 'title' }, { key: 'subtitle' },{ key: 'name' }]
        });

        expect(wrapper.vm.isFieldsetVisible({ id: 'fieldset_1'})).toBe(false);
    })

});