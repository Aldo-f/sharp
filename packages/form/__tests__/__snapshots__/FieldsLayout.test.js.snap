// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fields-layout can mount "fieldset" fields layout 1`] = `
<div class="GRID_MOCK">
	<fieldset>
		<legend class="SharpForm__label form-label">Fieldset</legend>
		<div class="card SharpForm__fieldset shadow-sm">
			<div class="card-body">
				<div class="FIELDS_LAYOUT_MOCK" layout="[object Object]">
					<div class="FIELD_MOCK" data="[object Object]"></div>
				</div>
			</div>
		</div>
	</fieldset>
</div>
`;

exports[`fields-layout can mount "hidden fieldset" fields layout 1`] = `
<div class="GRID_MOCK">
	<fieldset style="display: none;">
		<legend class="SharpForm__label form-label">Fieldset</legend>
		<div class="card SharpForm__fieldset shadow-sm">
			<div class="card-body">
				<div class="FIELDS_LAYOUT_MOCK" layout="[object Object]">
					<div class="FIELD_MOCK" data="[object Object]"></div>
				</div>
			</div>
		</div>
	</fieldset>
</div>
`;

exports[`fields-layout can mount fields layout 1`] = `
<div class="GRID_MOCK">
	<div class="FIELD_MOCK" data="[object Object]"></div>
</div>
`;
